### PR TITLE
cleanup, less redundancy between libretro & mame input

### PR DESCRIPTION
**Pull request by markwkidd**
This gets rid up a lot of redundant lines of code to determine the button that I wrote when implementing the control names for libretro frontends.

I am still testing this to make sure there not regressions, but it should not change any functionality. I'll merge once I make sure control descriptions are still making it into the right places.

### CODE REVIEW

**Comment by mahoneyt944** [comment]
Watch with calc_player_number. If we are changing this from it's index 0 to 5, to player number 1 to 6 then this function shouldn't subtract one anymore

**Review by markwkidd — COMMENTED** [review]
you did warn me to look at this index change closely SMH
I may had an idea that using the display number for the players would be better here but it has cost me in debugging time heh

**Comment by mahoneyt944** [comment]
@markwkidd hey I pushed a fix for calc_player_number. I try to avoid pushing to your wip to not mess up your commits.... But this one looked like it might get you all mixed up with the other changes your doing. You'll have to add it to your own copy.

**Comment by markwkidd** [comment]
Thanks that's very useful!

**Comment by markwkidd** [comment]
It's late, so I must stop for now. At the moment, this compiles, runs, and seems to describe controls correctly.

Therefore I rebased and force pushed in order to get a clean commit.

Is everything working right? I will test this more before I would say that 💯

**Comment by mahoneyt944** [comment]
Some oddities in the log:

controls "described as" repeats a lot? Loop  / break issue?
```
[MAME 2003+] Git Version  6f4a2ad
[MAME 2003+] Content lookup name: cclimber
[MAME 2003+] Driver index counter: 217. Matched game driver: cclimber
[MAME 2003+] Content identified as using dual joystick controls controls.
[MAME 2003+] Content identified as using dual joystick controls controls.
[MAME 2003+] Content identified as using dual joystick controls controls.
[MAME 2003+] Content identified as using dual joystick controls controls.
[MAME 2003+] Content identified as using dual joystick controls controls.
[MAME 2003+] Content identified as using dual joystick controls controls.
[MAME 2003+] Content identified as using dual joystick controls controls.
[MAME 2003+] Content identified as using dual joystick controls controls.
[MAME 2003+] Content identified as using dual joystick controls controls.
[MAME 2003+] Content identified as using dual joystick controls controls.
[MAME 2003+] Content identified as using dual joystick controls controls.
[MAME 2003+] Content identified as using dual joystick controls controls.
[MAME 2003+] Content identified as using dual joystick controls controls.
[MAME 2003+] Content identified as using dual joystick controls controls.
[MAME 2003+] Content identified as using dual joystick controls controls.
[MAME 2003+] Content identified as using dual joystick controls controls.
[MAME 2003+] Content identified as using 8-way joystick controls.
[MAME 2003+] Content identified by controls.c as having joysticks on axis with respect to the cabinet.
[MAME 2003+] Content identified as supporting 2 players with 2 distinct controls.
[MAME 2003+] Content identified as supporting 0 button controls.
[MAME 2003+] Content identified by controls.c as having mirrored multiplayer control labels.
```

This is just hard to follow, are we connecting unsupported controllers? Why?
```
[MAME 2003+] Preparing to connect input    in_port: 0    device: 1
[Environ]: SET_INPUT_DESCRIPTORS:
player_number: 1 | active_control_type[player_number - 1]: 1
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1002 | id: 2 | desc: Coin
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1003 | id: 3 | desc: Start
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1004 | id: 4 | desc: Up
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1005 | id: 5 | desc: Down
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1006 | id: 6 | desc: Left
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1007 | id: 7 | desc: Right
player_number: 2 | active_control_type[player_number - 1]: 0
[Environ]: SET_INPUT_DESCRIPTORS:
	RetroPad, Port 1, Button "Select" => "Coin"
	RetroPad, Port 1, Button "Start" => "Start"
	RetroPad, Port 1, Button "D-Pad Up" => "Up"
	RetroPad, Port 1, Button "D-Pad Down" => "Down"
	RetroPad, Port 1, Button "D-Pad Left" => "Left"
	RetroPad, Port 1, Button "D-Pad Right" => "Right"
[MAME 2003+] Preparing to connect input    in_port: 1    device: 0
[Environ]: SET_INPUT_DESCRIPTORS:
player_number: 1 | active_control_type[player_number - 1]: 1
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1002 | id: 2 | desc: Coin
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1003 | id: 3 | desc: Start
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1004 | id: 4 | desc: Up
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1005 | id: 5 | desc: Down
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1006 | id: 6 | desc: Left
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1007 | id: 7 | desc: Right
player_number: 2 | active_control_type[player_number - 1]: 0
[Environ]: SET_INPUT_DESCRIPTORS:
	RetroPad, Port 1, Button "Select" => "Coin"
	RetroPad, Port 1, Button "Start" => "Start"
	RetroPad, Port 1, Button "D-Pad Up" => "Up"
	RetroPad, Port 1, Button "D-Pad Down" => "Down"
	RetroPad, Port 1, Button "D-Pad Left" => "Left"
	RetroPad, Port 1, Button "D-Pad Right" => "Right"
[MAME 2003+] Preparing to connect input    in_port: 2    device: 0
[Environ]: SET_INPUT_DESCRIPTORS:
player_number: 1 | active_control_type[player_number - 1]: 1
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1002 | id: 2 | desc: Coin
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1003 | id: 3 | desc: Start
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1004 | id: 4 | desc: Up
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1005 | id: 5 | desc: Down
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1006 | id: 6 | desc: Left
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1007 | id: 7 | desc: Right
player_number: 2 | active_control_type[player_number - 1]: 0
[Environ]: SET_INPUT_DESCRIPTORS:
	RetroPad, Port 1, Button "Select" => "Coin"
	RetroPad, Port 1, Button "Start" => "Start"
	RetroPad, Port 1, Button "D-Pad Up" => "Up"
	RetroPad, Port 1, Button "D-Pad Down" => "Down"
	RetroPad, Port 1, Button "D-Pad Left" => "Left"
	RetroPad, Port 1, Button "D-Pad Right" => "Right"
[MAME 2003+] Preparing to connect input    in_port: 3    device: 0
[Environ]: SET_INPUT_DESCRIPTORS:
player_number: 1 | active_control_type[player_number - 1]: 1
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1002 | id: 2 | desc: Coin
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1003 | id: 3 | desc: Start
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1004 | id: 4 | desc: Up
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1005 | id: 5 | desc: Down
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1006 | id: 6 | desc: Left
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1007 | id: 7 | desc: Right
player_number: 2 | active_control_type[player_number - 1]: 0
[Environ]: SET_INPUT_DESCRIPTORS:
	RetroPad, Port 1, Button "Select" => "Coin"
	RetroPad, Port 1, Button "Start" => "Start"
	RetroPad, Port 1, Button "D-Pad Up" => "Up"
	RetroPad, Port 1, Button "D-Pad Down" => "Down"
	RetroPad, Port 1, Button "D-Pad Left" => "Left"
	RetroPad, Port 1, Button "D-Pad Right" => "Right"
[MAME 2003+] Preparing to connect input    in_port: 4    device: 0
[Environ]: SET_INPUT_DESCRIPTORS:
player_number: 1 | active_control_type[player_number - 1]: 1
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1002 | id: 2 | desc: Coin
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1003 | id: 3 | desc: Start
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1004 | id: 4 | desc: Up
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1005 | id: 5 | desc: Down
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1006 | id: 6 | desc: Left
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1007 | id: 7 | desc: Right
player_number: 2 | active_control_type[player_number - 1]: 0
[Environ]: SET_INPUT_DESCRIPTORS:
	RetroPad, Port 1, Button "Select" => "Coin"
	RetroPad, Port 1, Button "Start" => "Start"
	RetroPad, Port 1, Button "D-Pad Up" => "Up"
	RetroPad, Port 1, Button "D-Pad Down" => "Down"
	RetroPad, Port 1, Button "D-Pad Left" => "Left"
	RetroPad, Port 1, Button "D-Pad Right" => "Right"
[MAME 2003+] Preparing to connect input    in_port: 5    device: 0
[Environ]: SET_INPUT_DESCRIPTORS:
player_number: 1 | active_control_type[player_number - 1]: 1
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1002 | id: 2 | desc: Coin
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1003 | id: 3 | desc: Start
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1004 | id: 4 | desc: Up
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1005 | id: 5 | desc: Down
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1006 | id: 6 | desc: Left
[MAME 2003+] Describing controls for player_number: 1 | osd_code: 1007 | id: 7 | desc: Right
player_number: 2 | active_control_type[player_number - 1]: 0
[Environ]: SET_INPUT_DESCRIPTORS:
	RetroPad, Port 1, Button "Select" => "Coin"
	RetroPad, Port 1, Button "Start" => "Start"
	RetroPad, Port 1, Button "D-Pad Up" => "Up"
	RetroPad, Port 1, Button "D-Pad Down" => "Down"
	RetroPad, Port 1, Button "D-Pad Left" => "Left"
	RetroPad, Port 1, Button "D-Pad Right" => "Right"
```

**Comment by markwkidd** [comment]
About supported controllers - RetroArch, at least as of the time I implemented the controller layouts, did not like the core telling the frontend that there should only be controller descriptions for the number of players supported by a particular game.

What I would like to be able to do is send controller descriptions for Player 1 and Player 2, and Player 3 let's say for 3P game, and then send RETRO_DEVICE_NONE or NULL controllers for Player 4-6.

That didn't work the way I hoped in RetroArch, so what I did instead was send an altered version of Player 4-6 with the word "UNSUPPORTED" added to its name to show that for the particular game, that controller port is not in use.

**Comment by markwkidd** [comment]
I agree this seems like a time to reduce the amount of debug logging in those content flags, and maybe I can see some ways to simplify it.

The process is a little redundant but I think in terms of looping it is working as designed. It looks at each MAME input port for the game and compares it to certain attributes, then sets a flag. That seems like the simplest way to check each port for all the different indicators, and it only happens once.

There's no reason for it to spam the log, and this is another area where I may be able to replace some code with pre-existing MAME functions so we don't need as much here.

Anyway I'll put this on the list of areas to look over.

**Comment by mahoneyt944** [comment]
@markwkidd just flag each control type as the function is looping then check the flags at the end and log it once for each flag that was set.

**Comment by mahoneyt944** [comment]
Seems like options.content_flags already covers this to a degree. Maybe just check against these flags and log at the end.

**Comment by markwkidd** [comment]
Yeah I think you are right. I'll rewrite it to log everything at the end once I get a minute.

**Comment by markwkidd** [comment]
(The LGTM warning is logical, but doesn't apply to the line that it flagged because it's comparing a range)

I have pushed the changed to content flag logging, and played around a little bit with some games. I believe this is working.
